In [1]:
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt
from sklearn import datasets
from sklearn.tree import DecisionTreeClassifier
from sklearn import tree

In [3]:
url='https://raw.githubusercontent.com/Geerdata/DS/main/Datacoder/Arc.%20Modelo/WA_Fn-UseC_-Telco-Customer-Churn.csv'

df= pd.read_csv(url,sep=',')

df.head()

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges  TotalCharges Churn  
0           Electronic check          29.85         29.85    No  
1               Mailed check          56.95        1889.5    No  
2               Mailed check          53.85        108.15   Yes  
3  Bank transfer (automatic)          42.30       1840.75    No  
4           Electronic check          70.70        151.65   Yes  

[5 rows x 21 columns]

In [28]:
x = df[['InternetService', 'TotalCharges', 'Contract', 'PaymentMethod']].copy()
y = df[['Churn']].copy()

In [16]:
x.head()

InternetService TotalCharges        Contract              PaymentMethod
0             DSL        29.85  Month-to-month           Electronic check
1             DSL       1889.5        One year               Mailed check
2             DSL       108.15  Month-to-month               Mailed check
3             DSL      1840.75        One year  Bank transfer (automatic)
4     Fiber optic       151.65  Month-to-month           Electronic check

In [17]:
x_encoded = pd.get_dummies(x)
x_encoded.head()

InternetService_DSL  InternetService_Fiber optic  InternetService_No  \
0                 True                        False               False   
1                 True                        False               False   
2                 True                        False               False   
3                 True                        False               False   
4                False                         True               False   

   TotalCharges_   TotalCharges_100.2  TotalCharges_100.25  \
0           False               False                False   
1           False               False                False   
2           False               False                False   
3           False               False                False   
4           False               False                False   

   TotalCharges_100.35  TotalCharges_100.4  TotalCharges_100.8  \
0                False               False               False   
1                False               False               False   
2                False               False               False   
3                False               False               False   
4                False               False               False   

   TotalCharges_100.9  ...  TotalCharges_999.45  TotalCharges_999.8  \
0               False  ...                False               False   
1               False  ...                False               False   
2               False  ...                False               False   
3               False  ...                False               False   
4               False  ...                False               False   

   TotalCharges_999.9  Contract_Month-to-month  Contract_One year  \
0               False                     True              False   
1               False                    False               True   
2               False                     True              False   
3               False                    False               True   
4               False                     True              False   

   Contract_Two year  PaymentMethod_Bank transfer (automatic)  \
0              False                                    False   
1              False                                    False   
2              False                                    False   
3              False                                     True   
4              False                                    False   

   PaymentMethod_Credit card (automatic)  PaymentMethod_Electronic check  \
0                                  False                            True   
1                                  False                           False   
2                                  False                           False   
3                                  False                           False   
4                                  False                            True   

   PaymentMethod_Mailed check  
0                       False  
1                        True  
2                        True  
3                       False  
4                       False  

[5 rows x 6541 columns]

Ahora no va 0,1 sino que muestra true false

In [18]:
x_encoded.shape

(7043, 6541)

In [26]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(x_encoded, y, test_size=0.3, random_state=42)#test depende del tamaño del dataset en general, semilla de donde arranca

In [29]:
# Entrenar el arbol
from sklearn import tree
clf = tree.DecisionTreeClassifier()
clf = clf.fit(X_train, y_train)

In [30]:
pred=clf.predict(X_test)

In [32]:
from sklearn.metrics import classification_report
print(classification_report(y_true=y_test,y_pred=pred))

              precision    recall  f1-score   support

          No       0.81      0.89      0.85      1539
         Yes       0.61      0.45      0.52       574

    accuracy                           0.77      2113
   macro avg       0.71      0.67      0.69      2113
weighted avg       0.76      0.77      0.76      2113



De todo el muestreo se llega al 77% de acertividad, a nivel entrenamiento es bueno
- Precision: El objetivo principal es predecir los que se van a dar de baja, nuestro modelo tiene mas precision en las personas que no se van a dar de baja, con respecto a las que si se van a dar de baja.